_aicspylibczi/TypedImage.h

```diff
@@ -124,24 +124,21 @@ TypedImage<T>::loadImage(const std::shared_ptr<libCZI::IBitmapData>& bitmap_ptr_
   libCZI::ScopedBitmapLockerP lckScoped{ bitmap_ptr_.get() };
   // WARNING do not compute the end of the array by multiplying stride by
   // height, they are both uint32_t and you'll get an overflow for larger images
-  uint8_t* sEnd = static_cast<uint8_t*>(lckScoped.ptrDataRoi) + lckScoped.size;
-  if ((lckScoped.stride % size_.w == 0) && ( m_pixelType == libCZI::PixelType::Gray8 || m_pixelType == libCZI::PixelType::Gray16 ||
-      m_pixelType == libCZI::PixelType::Gray32Float) &&
-      samples_per_pixel_ != 3 ) {
+  if (lckScoped.stride % size_.w == 0 && lckScoped.stride / size_.w == sizeof(T)) {
+    // this is the vast majority of cases
     std::memcpy(m_array, lckScoped.ptrDataRoi, lckScoped.size);
-  }
-  else {
-    SourceRange<T> sourceRange(
-      samples_per_pixel_, static_cast<T*>(lckScoped.ptrDataRoi), (T*)(sEnd), lckScoped.stride, size_.w);
-    TargetRange<T> targetRange(samples_per_pixel_, size_.w, size_.h, m_array, m_array + length());
-    for (std::uint32_t h = 0; h < bitmap_ptr_->GetHeight(); ++h) {
-      pairedForEach(sourceRange.strideBegin(h),
-                    sourceRange.strideEnd(h),
-                    targetRange.strideBegin(h),
-                    [&](std::vector<T*> src_, std::vector<T*> tgt_) {
-                      pairedForEach(src_.begin(), src_.end(), tgt_.begin(), [&](T* s_, T* t_) { *t_ = *s_; });
-                    });
+  } else if (lckScoped.stride > size_.w) {
```

Comment:
is stride in bytes or pixels?  usually I expect stride to be bytes, but a "width" value to be pixels

Reply:
Stride is in bytes. The first condition Stride%w checks to see if it's an integer multiple. If it's not the else if is just to check that stride is larger than the width which it should be. If it's not then something weird is going on. IE compression etc which isn't being handled.

Reply:
I could be wrong but I think the proper conditions to check are whether (1) stride == width*bytesPerPixel, or (2) stride > width*bytesPerPixel.  if stride < width*bytesPerPixel then you have weirdness/error.  Doesn't that seem to read more like what is intended?

Reply:
You're right on point 2 but point 1 is equivalent. Likely Stride%width==0 should be sufficient to ensure they are in register.

Reply:
Modulo is not equivalent to equal:  if the stride were an integer multiple (2x)  the width, then your code would fall into the first memcpy, and would be wrong.  Of course, it would definitely be unexpected for that situation to happen, but you could write code to avoid it with certainty by using equality instead of modulo.  Because it's so unlikely to happen ,you can "get away" with the modulo.